File: programacao_dinamica/calcular_anagrama_dinamico.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <sys/resource.h>

#define MAX_PALAVRA 100
#define ALPHABET_SIZE 26
#define MAX_FAT 100

typedef struct {
    char *path;
    int qtd_palavras;
    int qtd_testes;
    int mostrar_resultado;
} args;

// =========================== VERSÃO DINÂMICA ===========================

void calcularFatoriais(long long *fatorial, int n) {
    fatorial[0] = 1;
    for (int i = 1; i <= n; i++) {
        fatorial[i] = fatorial[i - 1] * i;
    }
}

long long calcularAnagramasDinamico(const char *palavra) {
    int freq[ALPHABET_SIZE] = {0};
    int n = strlen(palavra);
    long long fatorial[MAX_FAT];
    calcularFatoriais(fatorial, n);

    for (int i = 0; i < n; i++) {
        char c = palavra[i];
        if (c >= 'A' && c <= 'Z') c += 32;
        if (c >= 'a' && c <= 'z') freq[c - 'a']++;
    }

    long long resultado = fatorial[n];
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        if (freq[i] > 1) resultado /= fatorial[freq[i]];
    }

    return resultado;
}

// =========================== FUNÇÃO PARA MEDIR MEMÓRIA ===========================

long get_mem_kb() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    return usage.ru_maxrss; // em KB
}

// =========================== ARGUMENTOS E UTILITÁRIOS ===========================

void ler_args(int argc, char *argv[], args *a) {
    if (argc < 5) {
        printf("Uso: %s <arquivo.txt> <qtd_palavras> <qtd_testes> <mostrar_resultado>\n", argv[0]);
        exit(1);
    }
    a->path = argv[1];
    a->qtd_palavras = atoi(argv[2]);
    a->qtd_testes = atoi(argv[3]);
    a->mostrar_resultado = atoi(argv[4]);
    if (a->qtd_palavras <= 0 || a->qtd_testes <= 0) {
        printf("Parâmetros inválidos!\n");
        exit(1);
    }
}

void ler_palavras(char **palavras, int qtd, const char *path) {
    FILE *f = fopen(path, "r");
    if (!f) {
        perror("Erro ao abrir o arquivo");
        exit(1);
    }

    char buffer[MAX_PALAVRA];
    int count = 0;

    while (fgets(buffer, sizeof(buffer), f) && count < qtd) {
        buffer[strcspn(buffer, "\n")] = '\0';  // Remove o '\n' do final
        palavras[count] = malloc(strlen(buffer) + 1);
        strcpy(palavras[count], buffer);
        count++;
    }

    fclose(f);

    if (count < qtd) {
        printf("⚠️  Aviso: arquivo contém apenas %d palavras, mas %d foram solicitadas.\n", count, qtd);
    }
}

void mostrar_palavras(char **palavras,int qtd){
    for(int i=0;i<qtd;i++){
        printf("%d - %s\n",i,palavras[i]);
    }
}

int main(int argc, char *argv[]) {
    args a;
    ler_args(argc, argv, &a);

    char **palavras = malloc(a.qtd_palavras * sizeof(char *));
    ler_palavras(palavras, a.qtd_palavras, a.path);
    // mostrar_palavras(palavras,a.qtd_palavras);

    double tempo_dyn = 0;
    long long res_dyn = 0;
    long memoria_dyn = 0;

    for (int i = 0; i < a.qtd_palavras; i++) {
        

        for (int t = 0; t < a.qtd_testes; t++) {
            long mem_before = get_mem_kb();
            clock_t inicio = clock();
            res_dyn = calcularAnagramasDinamico(palavras[i]);
            clock_t fim = clock();
            long mem_after = get_mem_kb();
            tempo_dyn += (double)(fim - inicio) / CLOCKS_PER_SEC;
            memoria_dyn += (mem_after - mem_before);
        }
        
        if (a.mostrar_resultado)
            printf("Palavra: %-20s | Rec: %.6fs, %ldKB |  Resultado: %lld\n", palavras[i], tempo_dyn, memoria_dyn, res_dyn);
    }
    
    tempo_dyn /= a.qtd_testes;
    memoria_dyn /= a.qtd_testes;

    printf("Tempo:%.6fs Memoria:%ldKB\n",tempo_dyn, memoria_dyn);

    for (int i = 0; i < a.qtd_palavras; i++)
        free(palavras[i]);
    free(palavras);

    return 0;
}

File: programacao_dinamica/calcular_anagrama_dinamico.py
import sys
import time
import tracemalloc
import gc
from collections import Counter

class Args:
    def __init__(self, path, qtd_testes, mostrar_resultado):
        self.path = path
        self.qtd_testes = qtd_testes
        self.mostrar_resultado = mostrar_resultado

# =========================== VERSÃO ITERATIVA ===========================

# Função iterativa para calcular fatorial
def fatorial_iter(n):
    resultado = 1
    for i in range(2, n+1):
        resultado *= i
    return resultado

# Função iterativa para contar frequências de letras
def contar_frequencias_iter(palavra):
    freq = Counter()
    for c in palavra:
        freq[c.lower()] += 1
    return freq

# Função iterativa para calcular o denominador da fórmula
def calcular_denominador_iter(freq_values):
    denominador = 1
    for count in freq_values:
        if count > 1:  # Otimização: não precisamos calcular fatorial de 1
            denominador *= fatorial_iter(count)
    return denominador

# Função principal para calcular anagramas de forma iterativa
def calcular_anagramas_iterativo(palavra):
    # Inicializa o contador de frequências
    freq = contar_frequencias_iter(palavra)
    
    # Calcula o numerador (fatorial do tamanho da palavra)
    numerador = fatorial_iter(len(palavra))
    
    # Calcula o denominador (produto dos fatoriais das frequências)
    denominador = calcular_denominador_iter(freq.values())
    
    # Retorna o resultado da fórmula
    return numerador // denominador

# =========================== FUNÇÃO PARA MEDIR MEMÓRIA E TEMPO ===========================

def measure_performance(func, *args):
    gc.collect()  # Força coleta de lixo antes da medição
    
    # Mede tempo
    inicio = time.time()
    
    # Mede memória
    tracemalloc.start()
    result = func(*args)
    current, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()
    
    # Calcula tempo
    fim = time.time()
    tempo = fim - inicio
    
    return result, tempo, peak / 1024  # Resultado, tempo em segundos, memória em KB

# =========================== ARGUMENTOS E UTILITÁRIOS ===========================

def ler_args(argc, argv):
    if argc < 4:
        print(f"Uso: {argv[0]} <arquivo.txt> <qtd_testes> <mostrar_resultado>")
        sys.exit(1)
    
    path = argv[1]
    qtd_testes = int(argv[2])
    mostrar_resultado = int(argv[3])
    return Args(path, qtd_testes, mostrar_resultado)

def ler_palavra(path):
    try:
        with open(path, 'r') as f:
            palavra = f.readline().strip()
            return palavra
    except FileNotFoundError:
        print(f"Erro: arquivo {path} não encontrado.")
        sys.exit(1)

# =========================== MAIN ===========================

def main():
    args = ler_args(len(sys.argv), sys.argv)
    palavra = ler_palavra(args.path)
    
    tempo_total = 0
    memoria_total = 0
    resultado = None
    
    for i in range(args.qtd_testes):
        res, tempo, memoria = measure_performance(calcular_anagramas_iterativo, palavra)
        tempo_total += tempo
        memoria_total += memoria
        resultado = res
    
    tempo_medio = tempo_total / args.qtd_testes
    memoria_media = memoria_total / args.qtd_testes
    
    if args.mostrar_resultado:
        print(f"Palavra: {palavra}")
        print(f"Quantidade de anagramas: {resultado}")
        print(f"Tempo médio: {tempo_medio:.6f}s")
        print(f"Memória média: {memoria_media:.2f} KB")
    
    print(f"Tempo:{tempo_medio:.6f}s Memoria:{memoria_media:.2f}KB QTD:{len(palavra)}")

if __name__ == "__main__":
    main()